F!! [Boost].UT test framework integration support

lp55 opened the pull request:
Hi,

   This is the initial implementation of the [Boost].UT test framework integration support, developed alongside @claremacrae 
    I see that I still have to build the tests for the approval tests of the integration with the test framework :) But please take a look to see if anything else is missing and if there are changes needed for what's been done.

claremacrae commented on ApprovalTests/integrations/ut/UTApprovals.h:
```diff
@@ -0,0 +1,80 @@
+                    boost::ut::reporter<boost::ut::printer>::on(test_begin);
+                }
+
+                auto on(boost::ut::events::test_run test_run) -> void { 
```
I wonder whether these one-liner methods that call the base class are required - it looks like it's calling the method of the same name in the base class, so could these be removed, for simplification?

No, I'm wrong - I commented out some of these in the project we experimented with yesterday, and got compiler errors - so they are required. :-)

claremacrae commented on ApprovalTests/namers/ApprovalTestNamer.h:
```diff
@@ -109,6 +109,11 @@ R"(* Welcome to Approval Tests.
 *     #define APPROVALS_DOCTEST
 *     #include "ApprovalTests.hpp"
 *
+* To do this in UT, add the following to your main.cpp:
```
This is brilliant - thank you for updating this.

Trivia for info: this changing is why the CI tests failed - there is an approval test (in dockets tests, I think) that approves the content of this text - so that we can include it in the documentation. Someone will at some point need to re-run that test and approve the new output, then the tests will pass again.

claremacrae commented on doc/UsingUT.md:
```diff
@@ -0,0 +1,99 @@
+<!--
```
Thanks ever so much for updating the documentation...

Did you edit and create mdsource/.source.md files - if so, please could you add them to the PR?

If not, then they are explained here: https://github.com/approvals/ApprovalTests.cpp/blob/master/doc/Contributing.md#documentation

Note that I've just pushed some updates to master, to the generated documentation, with the latest mdsnippets tool - apologies if this creates conflicts for you.

Let me know if you have any questions about this... if it saves you time, it's fine for you to leave the docs in these normally-generated files for now, and I can move the text around before merging...

lp55 replied:
Hi, the only new file I created for the documentation is this doc/UsingUT.md, which I copy/pasted and then changed from the doc/UsingDoctest.md. I didn't see the doc\mdsource subdirectory. I just added the file now.

claremacrae replied:
Thanks for this... We will need to get your other markdown edits in the master source.md files, so that they don't get overwritten next time the docs are generated... If you'd like me to do this after the merge, then please resolve this conversation....

claremacrae commented on doc/UsingUT.md:
```diff
@@ -0,0 +1,99 @@
+<sup>[snippet source](/tests/UT_Tests/main.cpp#L10-L19) / [anchor](#snippet-ut_main_usage)</sup>
+<!-- endsnippet -->
+
+In thw following example, two instances of ApprovalTests are called inside the same test:
```
Trivial typo: `thw`

claremacrae commented on doc/UsingUT.md:
```diff
@@ -0,0 +1,99 @@
+
+Bellow is an example of a call to an approval test inside a \[Boost\].UT test:
+
+<!-- snippet: ut_main_usage -->
```
If you would like to create example code specifically for documentation, feel free to add a new sub-directory to examples/

lp55 replied:
I think the documentation snippets for now will suffice. What I liked about UT is that it's so simple to use (and doesn't have macros). If more questions arise I can make a proper example project then.

claremacrae commented on ApprovalTests/integrations/ut/UTApprovals.h:
```diff
@@ -0,0 +1,80 @@
+#include "ApprovalTests/namers/ApprovalTestNamer.h"
+
+// <SingleHpp unalterable>
+#ifdef APPROVALS_UT
```
This is all great - thank you!

I've been debating between APPROVALS_UT and  APPROVALS_BOOST_UT.

I don't have a strong opinion, but will at least write down my thoughts, to see what you think...

If we add support for boost.test, it will presumably have to be called APPROVALS_BOOST_TEST, as APPROVALS_TEST would be too ambiguous.

So that might suggest APPROVALS_BOOST_UT, for consistency.

But then there is the possibility that UT doesn't get accepted to boost - in which case the shorter name would end up better...

In general I tend to prefer longer names, if they are clearer....

What do you think?

lp55 replied:
I was in doubt about that as well. As all the things in this boost-experimental were never officially approved into boost (even although they are high quality) I presumed this one won't either. So for the defines and files names I just stuff with UT. On the documentation text I used [Boost] UT as the brackets indicate the non official nature of the boost keyword.

claremacrae replied:
Good points. Thanks. I agree with all that, so am marking this as Resolved.

claremacrae commented on tests/UT_Tests/main.cpp:
```diff
@@ -0,0 +1,42 @@
+
+	"Approval"_test = []() {
+        expect(nothrow([] {
+            Approvals::verify("Approval Tests can verify text via the golder master method");
```
Minor typo - it's golden

claremacrae commented:
This is great work - thank you very much indeed... 

Some extra comments apart from the code review:

* This is definitely on the right track, and will be really useful.
* Thanks for using the `F!!` prefix! 
* What's the behaviour now when an Approval Test verify fails? Is the text of the exception written out? If not, can this be made to happen please, so users will know how to deal with test failures.
* For writing tests of the naming, this one might give you some ideas: `tests/DocTest_Tests/namers/DocTestNamerTests.cpp`
* At some point it would be good to have a ut release header in third_party - note the dance that we do with adding the release file with its version name, and then a one-line file that wraps around - see `third_party/doctest/doctest.2.3.5.h` for an example...

Thanks again!

Clare

lp55 commented:
One thing, I caught a bug when compiling UT inside ApprovalTests. It was due to a min macro MSVC adds. I made a pull request on UT: https://github.com/boost-experimental/ut/pull/235
The developer seems very responsive, so this should be patched soon, and I'll up the minimum version for UT on the documentation afterwards.

lp55 commented:
As for adding UT to the third_party, I recommend waiting for it to stabilize a bit more first. It's fairly new and still being heavily developed. You would have an obsolete version very quickly.

claremacrae commented:
> As for adding UT to the third_party, I recommend waiting for it to stabilize a bit more first. It's fairly new and still being heavily developed. You would have an obsolete version very quickly.

I understand...

I should have explained why... As we maintain Approval Tests, we would like to know that we haven't broken any of our integrations, which means we need to run all our tests, including those of all our integrations, in our CI...

In the longer run, I know that we could improve our CMake files a bit, and download dependencies - but for now, the way we do this is via third_party...

So having an out-of-date file in third_party is preferable to not running some of our tests...

claremacrae commented:
Luiz, I should have said, I'm really happy to pair again on any of the remaining bits of this. If that would help, let's set it up by email...

lp55 commented:
As for the behaviour when an Approval Test verify fails, the user can just do something like this:

```cpp
expect(nothrow([] { 
    auto section = NamerFactory::appendToOutputFilename("test 1");
    Approvals::verify("Approval Tests can verify text via the golder master method"); 
})) << "Approval test error";
```

I can update the docs with that. Or do you have something else in mind?

lp55 commented:
I've added more tests. Tomorrow I'll improve the documentation and it should be ok to merge.

lp55 commented:
Hi, I've fixed a check on cmakefiles to only compile ut integration with the supported compilers and hand fixed the UsingUT.md documentation (I don't have the software to generate the mds from the source). I fixed the snippets to use from real code now. Should be good to merge. If there's anything else need, just ask.

claremacrae commented:
The tests all pass for me, so I'm accepting this - thanks very much @lp55!

Afterwards, I'll pull in the latest ut header, with exception and source_location fixes - and then remove the try/catch block from the tests..

claremacrae commented:
Hi @lp55 - it's all looking good!

I added a "Minimum C++ Version" column to this table: https://github.com/approvals/ApprovalTests.cpp/blob/master/doc/GettingStarted.md#choosing-a-testing-framework

I said C++17 for UT, but then I noticed it now says `CMAKE_CXX_STANDARD` 20 in the CMakeLists.txt for the tests...

Should I change the table to say 20, or can the CMake file be changed back to 17?

Thanks!

lp55 commented:
Hi @claremacrae,

Thanks for accepting the pull request!
Actually is better to say C++20, because source_location is actually a C++20 feature, and for ApprovalTest it's required, event if the UT framework itself can work with C++17.

claremacrae commented:
Thanks @lp55 - I've updated the table to C++20 and added a footnote explaining why:

https://github.com/approvals/ApprovalTests.cpp/blob/master/doc/GettingStarted.md#choosing-a-testing-framework
